### ISSUE TITLE
CSS color decorators shown erroneously in a number of places in tsx file

### ISSUE THREAD

**Issue by afeinman** [styled-components/vscode-styled-components#296]
Background:
* Extension-bisect identified vscode-styled-components as the culprit.
* vscode-styled-components version: 1.6.1
* VS Code version: Code 1.56.2 (054a9295330880ed74ceaedda236253b4f39a335, 2021-05-12T17:44:30.902Z)
* OS version: Darwin x64 20.4.0

<img width="633" alt="highlights" src="https://user-images.githubusercontent.com/561248/119409477-5f5d8900-bcb5-11eb-9afa-76fda2167af7.png">

NOTE: File type must be Typescript React, not typescript.
We have a map which stores our GUI colors by name, e.g.

```const Colors = { "Our Blue": "#400080", "Dark Yellow": "#884400" };```

1. When accessing this map, VSCode incorrectly offers a color picker for the color "Blue" (#0000FF) at ⌧ below.

```
const pageColors = {
   bgColor: Colors["Our ⌧Blue"],
}
```

This appears to require the space in the key name ("Our Blue") to trick the parser.

2. key names that are colors will also get an erroneous picker. Changing this key would lead to bugs.

```
  ⌧black: Colors["Black"]
```

3. Color pickers show in some other places when the line also contains other color-ish strings:

```
const Colors2 = { "Our Blue": "⌧#400080", "Dark Yellow": "#884400" };
```

A picker appears before #400080 but not before #884400.

**Comment by jasonwilliams** [comment]
Thanks for raising this @afeinman 
So there are 2 problems here (which you’ve spotted)

- The first is that we match on any colour name regardless of where it is in the file. I think I may remove this feature and only match the other formats . Matching the name yields too many false positives, like the one in your example. I also don’t believe that many developers use css colour names anyway.
- Secondly the Regex only matches once per line. So that’s why yellow doesn’t match in your example. This is fixable and we can track that here

**Comment by afeinman** [comment]
Hmm. For the first problem, perhaps tighten up the colour name matching to require it being inside a string constant (assuming you can access the AST) and for the entire string to match a color name?

The second one may result naturally from moving to AST-based matching rather than regex.

**Comment by jasonwilliams** [comment]
> Hmm. For the first problem, perhaps tighten up the colour name matching to require it being inside a string constant (assuming you can access the AST) and for the entire string to match a color name?

There’s no way to know that unless it’s some sort of AST based match which I’m not willing to do at this point. I think I’ll have it either follow a : or a comma

**Comment by jasonwilliams** [comment]
@afeinman this has been fixed in https://github.com/styled-components/vscode-styled-components/commit/bd8cec3dda4a1e90d359087650f888c0d3033ac8 should be released as part of v1.6.4
